### PR TITLE
fix(seed): fix console.log in basic JavaScript challenges

### PR DESCRIPTION
**Pull request by QuincyLarson**
This fixes the issue of the output console on the JavaScript challenges not properly returning the output of functions. This attempts to fix the "tail" section of challenges.

Currently, ~~there are several broken tests,~~ and the ones that do pass have net been fully QA'd to verify that they properly return output.

Edit (raisedadead): 
- [x] Broken tests are fixed.
- [x] Refers part of #9363.
- [x] Tested locally.

### CODE REVIEW

**Comment by QuincyLarson** [comment]
@BerkeleyTrue This is the tail issue that I started a couple months ago. I had some trouble with it. Since you understand what's going on under the hood better than I do, you're in a better position to fix these breaking tests and ship this feature.

**Comment by raisedadead** [comment]
@QuincyLarson does your PR target #9363 ?

**Comment by raisedadead** [comment]
@QuincyLarson I have resolved some conflicts with this Pull Request. Will test and check locally, let me know if you need any changes.

**Comment by raisedadead** [comment]
@QuincyLarson I have fixed the broken CI and the tails for all of the challenges that failed. Please have a look.

@FreeCodeCamp/moderators it would be awesome if any of you could help in QA and testing locally as well.

**Review by BerkeleyTrue — APPROVED** [review]

File: seed/challenges/01-front-end-development-certification/basic-javascript.json [hunk 1]
```diff
@@ -85,7 +85,7 @@
         ""
       ],
       "tail": [
-        "if(typeof myName !== \"undefined\"){(function(v){return v;})(myName);}"
+        "if(typeof myName !== \"undefined\"){ console.log(myName); }"
```
Comment (erictleung) [inline seed/challenges/01-front-end-development-certification/basic-javascript.json]: Is this tail code necessary? This challenge, "Declare JavaScript Variables," just asks to declare a variable, without setting it equal to anything.

**Comment by erictleung** [comment]
I can't get the challenges to output anything to either the page's console or the browser's console. 

Here's "Initializing Variables with the Assignment Operator" and it has the new tail code of

```js
"if(typeof a !== 'undefined') { console.log(\"a = \" + a); } else { console.log('a is undefined'); }"
```

image

Or is the issue I'm bringing up a different issue?

File: seed/challenges/01-front-end-development-certification/basic-javascript.json (hunk 1 above)
Comment (BerkeleyTrue) [inline seed/challenges/01-front-end-development-certification/basic-javascript.json]: Maybe it should just log when myName is defined?

Comment (raisedadead) [inline seed/challenges/01-front-end-development-certification/basic-javascript.json]: Oh yes.

**Comment by raisedadead** [comment]
Tagging as blocked, I'll will check this ASAP.

**Comment by raisedadead** [comment]
image

@erictleung it seems to work for me, I tested by completely purging the database, doing a npm install, and seeding from scratch.

**Comment by raisedadead** [comment]
This is ready for QA, I have updated log for the review comment.

**Comment by raisedadead** [comment]
@freeCodeCamp/moderators Bump for a QA. I will update the challenges further in separate PR/s, this one is on Quincy's fork, so did a basic clean up / rebase in addition to previous fixes.

**Comment by hallaathrad** [comment]
Can't do today, but if nobody has checked it by tomorrow morning (before my local session), I'll give it a try.

**Review by systimotic — COMMENTED** [review]
So I scrolled trough the code, and left a few comments.
Most of it seems 👌 
I checked a few random challenges. I checked:
- initializing-variables-with-the-assignment-operator
- increment-a-number-with-javascript (could not pass the challenge, but logging worked)
- compound-assignment-with-augmented-addition
- escape-sequences-in-strings
- use-bracket-notation-to-find-the-first-character-in-a-string

These all worked fine. 👍 

❌ declare-string-variables did not work
Edit: left comment for 👆 
❌ use-the-parseint-function did not show anything
Edit: seems like this was part of the expansion, this is done by function call, not by tail. I guess this may be handled in a future PR

Notes:
- I also had to drop the challenges collection to get this to work
- In the future, we may want to look at using the native logging instead of stringifying for some of the concatenated output. That's not important for this PR though.
- It seems like there are some more inconsistencies with IIFE's. They were there before this PR though, so I think it's fine to leave them as is for now. The ones I commented on should probably be changed, the ones I left should be looked at later.

File: seed/challenges/02-javascript-algorithms-and-data-structures/basic-javascript.json [hunk 2]
```diff
@@ -1038,7 +1045,7 @@
         ""
       ],
       "tail": [
-        "if(typeof myFirstName !== \"undefined\" && typeof myLastName !== \"undefined\"){(function(){return myFirstName + ', ' + myLastName;})();}"
+        "if(typeof myFirstName !== \"undefined\" && typeof myLastName !== \"undefined\"){ console.log('myFirstName + ', ' + myLastName'); }"
```
Comment (systimotic) [inline seed/challenges/02-javascript-algorithms-and-data-structures/basic-javascript.json]: Ah, seems like the reason for this not working is obvious, the extra quotes shouldn't be there.

File: seed/challenges/02-javascript-algorithms-and-data-structures/basic-javascript.json [hunk 3]
```diff
@@ -4779,7 +4792,7 @@
     },
     {
       "id": "56533eb9ac21ba0edf2244cc",
-      "title": "Accessing Nested Objects",
+      "title": "Access Nested Objects",
```
Comment (raisedadead) [inline seed/challenges/02-javascript-algorithms-and-data-structures/basic-javascript.json]: I don't remember doing this, maybe it crept in over cycles of rebasing, thanks for catching, I will update.

**Comment by raisedadead** [comment]
@systimotic thanks a lot for the review. I can see that with a rebase, this pull request is causing more troubles than without. Also it has loads of nasty commits and I believe it's best I close this and start of clean. @QuincyLarson Sorry that some of your work will be lost in the process.

Which brings me to a good point that you raised we should look at the `uncapture` and `capture`. These are not so clean and are as a result are logging in things twice.

And I must say that I had brought in this kind of a monkey-patch when we did not have any mechanism of logging at all.

Thanks everyone for your time on this, but its best we close and look at this from a fresh branch of staging.

**Pull request closed without merging**